src/main/webapp/js/map-loader.js

```diff
@@ -1,38 +1,133 @@
+let editMarker;
+let map;
+let displayMarkers = {};
  /**
  * Adds a new marker with a toggable info window
- * @param {Map} map 
  * @param {Number} lat 
  * @param {Number} lng 
- * @param {String} title 
  * @param {String} description 
  */
-function addLandmark(map, lat, lng, title, description){
+function addDisplayMarker(lat, lng, description){
   const marker = new google.maps.Marker({
     position: { lat: lat, lng: lng },
-    map: map,
-    title: title
+    map: map
   })
   const infoWindow = new google.maps.InfoWindow({
-    content: description
+    content: buildDeletableMarker( lat, lng, description)
   })
   marker.addListener('click', () => {
     infoWindow.open(map, marker)
   })
+
+  // Uses a dictionary with both lat and lng for faster lookup
+  if(!displayMarkers[lat]) displayMarkers[lat] = {};
+  displayMarkers[lat][lng] = marker;
 }
 
 /**
  * Main function
  */
 function createMap(){
-  const map = new google.maps.Map(document.getElementById('map'), {
+  map = new google.maps.Map(document.getElementById('map'), {
     center: {lat: 37.764419, lng: -122.445911},
     zoom: 12
   });
+  map.addListener('click', (event) => {
+    createMarkerForEdit(event.latLng.lat(), event.latLng.lng());
+  })
 
-  addLandmark(map, 37.819021, -122.477946, "Golden Gate Bridge", "This is the Golden Gate Bridge")
-  addLandmark(map, 37.802122, -122.418805, "Lombard Street", "This is Lombard St.")
-  addLandmark(map, 37.811039, -122.415175, "Fishermans Wharf", "This is Fishermans Wharf")
+  fetchMarkers();
+}
 
+ /** Fetches markers from the backend and adds them to the map. */
+function fetchMarkers(){
+  fetch('/markers').then((response) => {
+    return response.json();
+  }).then((markers) => {
+    markers.forEach((marker) => {
+     addDisplayMarker(marker.lat, marker.lng, marker.content)
+    });
+  });
+}
 
+/** Creates a marker that shows a textbox the user can edit. */
+function createMarkerForEdit(lat, lng){
+  // If we're already showing an editable marker, then remove it.
+  if(editMarker){
+   editMarker.setMap(null);
+  }
+  editMarker = new google.maps.Marker({
+    position: {lat: lat, lng: lng},
+    map: map
+  });
+  const infoWindow = new google.maps.InfoWindow({
+    content: buildInfoWindowInput(lat,lng)
+  });
+  // When the user closes the editable info window, remove the marker.
+  google.maps.event.addListener(infoWindow, 'closeclick', () => {
+    editMarker.setMap(null);
+  });
+  infoWindow.open(map, editMarker);
+ }
+
+/** Builds and returns HTML elements that show an editable textbox and a submit button. */
+function buildInfoWindowInput(lat, lng){
+  const textBox = document.createElement('textarea');
```

Comment:
Hmm... prefer to have all of this in a html and just render it, instead of building ui elements in javascript. But optional and up to you

Reply:
How would that work? I understand creating a new HTML with the template, but how would I reference it in this script?

Reply:
You can reference is by calling getClassName or getElementById, but I will leave this as optional.

Reply:
I am thinking this will be useful when displaying the images for our prototype... Will totally take it into account